File: Arrays/49-Group-Anagrams/49.cpp
#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<string>> group_anagrams(vector<string> &strs) {
    unordered_map<string, vector<string>> map;

    for (const auto &str : strs) {
      string sorted = str;
      sort(sorted.begin(), sorted.end());

      // Or map.find(key) == map.end();
      if (map.empty())
        map[sorted] = {};

      map[sorted].push_back(str);
    }

    vector<vector<string>> groups;

    for (const auto &group : map)
      groups.push_back(group.second);

    return groups;
  };
};

File: Arrays/49-Group-Anagrams/49.ts
function groupAnagrams(strs: string[]): string[][] {
  const hash: { [key: string]: string[] } = {};

  for (const str of strs) {
    const sorted = str.split('').sort().join();
    if (!hash[sorted]) hash[sorted] = [];
    hash[sorted].push(str);
  }

  return Object.values(hash);
}
